### ISSUE TITLE
Migrate to Protobuf 3 to Increase Language Compatibility

### ISSUE THREAD

**Issue by turbo** [bloomberg/comdb2#832]
Some languages, most notably C#, do not support proto2. The proto files for comdb2 use several specific features which have been deprecated in proto3, like optional values, defaults, required values, non-base-0 enums.

So, this is a feature request to migrate to proto3. The protobuf interface is an easy way to talk to the database and migrating will path the way for more adoption, e.g. a proper database driver for C#.

**Comment by riverszhang89** [comment]
> Some languages, most notably C#, do not support proto2.

To be more accurate, Protobuf 2.x do not support code generation in C#. To work it around, one could use [the legacy protobuf port from google](https://github.com/jskeet/protobuf-csharp-port).

Migration to Protobuf 3 would actually break compatibility. There are a lot of popular projects that still use [Protobuf 2.x](https://github.com/apache/spark/blob/master/pom.xml#L121), or even [Protobuf 1.x](https://github.com/apache/activemq/blob/master/pom.xml#L41). The Protobuf developers themselves [advise against it](https://github.com/google/protobuf/releases/tag/v3.0.0) as well:

> We recommend that new Protocol Buffers users use proto3. However, we do not
generally recommend that existing users migrate from proto2 from proto3 due
to API incompatibility, and we will continue to support proto2 for a long
time.

**Comment by mponomar** [comment]
The wire protocol is the same between 2 and 3 AFAIK.  I had a POC driver for another language that used proto3 with no database changes.  As a compromise we can provide proto3 files for writing drivers  We can continue to use 2 internally.

**Comment by riverszhang89** [comment]
@mponomar Protobuf 3 understands both proto2 and proto3 syntax, as stated [here](https://github.com/google/protobuf/releases/tag/v3.0.0):

```
A new notion "syntax" is introduced to specify whether a .proto file
uses proto2 or proto3:

// foo.proto
syntax = "proto3";
message Bar {...}

If omitted, the protocol buffer compiler generates a warning and "proto2" is
used as the default. This warning will be turned into an error in a future
release.
```

I would not assume the wire protocol is the same between 2 and 3 based on the statement above. I think we need to do is explicit add `syntax="proto2"` to our proto files to avoid the error in the future.

@turbo You could just use Protobuf 3 to compile Comdb2 proto files without any issues. The legacy C# port has become part of Protobuf 3 core now.

**Comment by riverszhang89** [comment]
Okay we already have `syntax="proto2"` in every proto file.

**Comment by riverszhang89** [comment]
@turbo I assume by converting the proto files to 3, both client and server speak in 3 then? Could you please build a proto 2 cdb2sql and see if the cdb2sql can talk with proto 3 Comdb2 server?

**Comment by turbo** [comment]
@riverszhang89 Well, the protocol *is* largely wire-compatible. So by "converting" I mean stripping the unsupported keywords from the proto def, as well as changing the enums. Default values are supplied by the client (I see the JDBC driver already does this).

> You could just use Protobuf 3 to compile Comdb2 proto files without any issues. The legacy C# port has become part of Protobuf 3 core now.

The protoc compiler will reject protos marked as proto2. It's true that Skeet's version was merged upstream, but it has been rewritten to only support proto3:

```text
$ protoc --csharp_out=. sqlresponse.proto 
--csharp_out: sqlresponse.proto: C# code generation only supports proto3 syntax
```

I'm currently trying two things (in C# that is):

- Manually convert the protos to proto3 and use the Google implementation
- Using protogen and protobuf-net

**Comment by turbo** [comment]
See also https://mgravell.github.io/protobuf-net/version#the-binary-protocol:

> The binary ProtoBuf protocol has not changed at all in the time in which it has been available open-source; protobuf-net provides an implementation of the binary protocol, and as such it is not currently directly meaningful to ask what version of the binary protocol is supported: there is only one version, and it is supported.

Using protobuf-net is probably the best way, since it requires no modifications to the proto2 files of Comdb2. `protogen` converts the proto2 files to class definitions and gives an idiomatic API.

So as far as I'm concerned (C#), this is resolved. It might still be valuable to supply proto3 defs, so I'll leave this issue open for now.

**Comment by riverszhang89** [comment]
> It's true that Skeet's version was merged upstream, but it has been rewritten to only support proto3.

Thank you for pointing it out. It'd be ideal if we could make our proto files understood by both protobuf 2 and 3 without introducing incompatibility with current proto files.

Off topic: Are you working on a native C# driver for Comdb2?

**Comment by turbo** [comment]
> Off topic: Are you working on a native C# driver for Comdb2?

Yes. We use Comdb2 in production for some analysis workloads and our whole stack is C#. I find the "Writing language bindings" page helpful, but man, the python code on there is full of bugs :)

But so far, using protobug-net + protogen, it works just fine. I'm writing down all the gotchas I experience along the way, try to reproduce them and submit a PR/Issue when I'm confident that it's not my fault :+1:

**Comment by mponomar** [comment]
You may also be interested in the [python API](https://github.com/bloomberg/python-comdb2).
That wraps the C api instead of implementing the protocol.  Thanks for the feedback!

**Comment by turbo** [comment]
Yeah, I've looked at that. I found the language binding page and the JDBC driver source code to be most helpful. Once you get the overall flow of data and query/response layout, it's quite easy. The hardest part is getting the type conversion (mostly of the datetime types) correct.
